### ISSUE TITLE
Running segmentation on GeoBench NeonTree dataset

### ISSUE THREAD

**Issue by nobeldang** [Panopticon-FM/panopticon#1]
Hi thanks for sharing your impressive work. I am trying to run the Panopticon model on the GeoBench's Neon Tree dataset for the segmentation purpose.
Can you provide the running script for that? I can not seem to find relevant config file anywhere under dinov2/configs 

Thank You

**Comment by LeWaldm** [comment]
Hi, 

thank you for your interest in Panopticon. 

We are going to release the code we used to evaluate Panopticon within the next 10 days. Alternatively, you could use the code snippet provided in the [readme](https://github.com/Panopticon-FM/panopticon?tab=readme-ov-file#using-panopticon) to extract representations. However, you would need to write your own training loop & segmentation head. 

I will ping you here once we have released the evaluation code.

**Comment by JohnMBrandt** [comment]
I would also be interested in seeing the config files for the geo bench evaluations! I have been trying to replicate the reported numbers with:
- UperNet head with auxiliary FCNHead on p3 modeled after mmseg
- Feature2Pyramid neck modeled after mmseg
- large batch size (128+)
- grid search over learning rate
- default loss weights, initialization methods, feature sizes, normalization parameters, and dropout used in mmseg Upernet

But have been unable to replicate even the baseline dinov2 numbers for most of the evaluation datasets, often ending up lower by ~0.05 mIoU

**Comment by LeWaldm** [comment]
Hi @JohnMBrandt , thank you for the detailed analysis. I will double check the results of Panopticon and DINOv2 in our evaluation codebase. We will publish the evaluation codebase on Monday EOD.

**Comment by nobeldang** [comment]
Hi thanks for the efforts and replying. Please keep us updated on this thread about when you will push the configs file and related code.

**Comment by LeWaldm** [comment]
Hi both, 

thank you for your patience. We just made our evaluation codebase [Geobreeze](https://github.com/geobreeze/geobreeze/tree/main/geobreeze) public!

Geobreeze is a refactor of our original evaluation codebase that generated the results reported in the paper. We did not refactor all bash scripts. If you have any questions, feel free to reach out to us or look into the original (unpolished) [code](https://github.com/ando-shah/fm-playground).

Re:
- @nobeldang please find the script for neontree [here](https://github.com/geobreeze/geobreeze/blob/main/scripts/segm.sh). Note that we are only using the RGB channels of neontree.
- @JohnMBrandt Thank you for your detailed comments. The logic for our segmentation training is mostly [here](https://github.com/geobreeze/geobreeze/blob/6018b29b31e5335ff8a6f1a96020fe5c596ca000/geobreeze/engine/lightning_task.py#L216). I double-checked our code and I might have found a bug. I will come back to you in the next couple of days. 

If you have any questions, don't hesitate to reach out!

**Comment by JohnMBrandt** [comment]
Thanks for following up on this! I'm curious to see what you come up with in the segmentation training debugging. I have trained (using plain pytorch), the following models to try to recreate some of your numbers. The models use the same hyperparameters as you do:

-Auxiliary FCN on p3 with 0.1 dropout
- UperNet decoder with: 512 channels, 0.1 dropout
- Feat2Pyramid neck with rescales=[4, 2, 1, 0.5]
- 50 epochs, no LR warmup, 128 batch size, 224 training size, AdamW optimizer with 0.01 WD and grid search on LR


With a frozen DiNOv2-Large, just using the RGB bands, I am able to get 0.746 mIoU on `m-chesapeake`, while your paper reports 0.785.

However, this discrepancy doesn't exist for Privthi 2.0 -- they report 0.695 on `m-chesapeake` and I am able to recreate that with 0.693.

**Comment by JohnMBrandt** [comment]
One final note that suggests something odd with `geobreeze`'s calculations -- your paper's reported results for Galileo are rather inconsistent to Galileo's own reported results. The segmentation results in particular are 7-10 mIoU higher while the classification results are sometimes the same and sometimes different.

| Result 1 | m-Cashew-Plant | m-SA-Crop-Type | m-So2Sat | m-Brick-Kiln | 
|----------|----------|----------|----------|----------|
| Galileo-Reported by authors    |   33.0       |      30.1    |    54.8      |     98.7     |         
| Galileo - Reported by panopticon  |     40.4     |      39.5    |     54.2     |     93.1     |

**Comment by JohnMBrandt** [comment]
Not quite sure what the bug is geobreeze, but the results I am getting are all over the place. The models seem to train properly.  But I get these numbers for m-chesapeake with dinov2 which appear much too high.


```
lr=5e-06_bsz=128_e=50
         metric       val best_classifier
0     test/loss  0.814420        lr=1e-05
1  test/F1Score  0.856779        lr=1e-05
2  test/jaccard  0.749738        lr=1e-05
3     test/mIoU  0.874126        lr=1e-05
lr=5e-05_bsz=128_e=50
         metric       val best_classifier
0     test/loss  0.529295       lr=0.0001
1  test/F1Score  0.872238       lr=0.0001
2  test/jaccard  0.773730       lr=0.0001
3     test/mIoU  0.889195       lr=0.0001
lr=0.0005_bsz=128_e=50
         metric       val best_classifier
0     test/loss  0.526494        lr=0.001
1  test/F1Score  0.874306        lr=0.001
2  test/jaccard  0.777035        lr=0.001
3     test/mIoU  0.891632        lr=0.001
```


I modified the lightning task's test step to also save the inference predictions, so that I could calculate the mIoU manually. 


```
def test_step(self, batch, batch_idx):

        x_dict, targets = batch
        logits, aux_logits = self.forward(x_dict)
        preds = torch.argmax(logits, dim=1)   # (B, H, W)
        loss = self.loss((logits, aux_logits), targets)
        self.log_metrics(logits, targets, loss, prefix="test", apply_argmax=True)
   
        for fn, pred_mask, target_mask in zip(
            x_dict['fn'],
            preds.cpu().numpy(),
            targets.cpu().numpy()
        ):
            idx_str = f"{fn:06d}"
            pred_name   = idx_str + "_pred.png"
            target_name = idx_str + "_target.png"
    
            pred_path   = os.path.join(self.pred_dir, pred_name)
            target_path = os.path.join(self.pred_dir, target_name)
    
            # save as 8-bit grayscale PNG
            Image.fromarray(pred_mask.astype(np.uint8)).save(pred_path)
            Image.fromarray(target_mask.astype(np.uint8)).save(target_path)
```

I then manually loaded all the predictions and targets, created a confusion matrix, and calculated mIoU, getting `0.6985469985512468` for `lr=0.0005_bsz=128_e=50`, which seems much more likely. So likely your model training code works fine but there is a bug when the mIoU is calculated.

**Comment by LeWaldm** [comment]
Thank you the additional details @JohnMBrandt!

I think you are spot on with your analysis. From what I gathered yesterday, it seems like we have calculated the micro jaccard metric and reported it as mIoU, where mIoU is actually macro jaccard.

I will recompute the results with the correct metrics and report back. We are at CVPR over the week, I hope to come back to you at the end of the week (if you are also there lmk & let's meet up!).

2 more thoughts:
- you can easily add other torchmetrics to be computed [here](https://github.com/geobreeze/geobreeze/blob/main/geobreeze/config/data/task/segmentation.yaml). See the [build_metrics](https://github.com/geobreeze/geobreeze/blob/main/geobreeze/engine/accelerated/utils/metrics.py) for details 
- I corrected the bug while refactoring, the actual evaluation code was [this](https://github.com/ando-shah/fm-playground/blob/ecfa7b8c04f28f62ec01a4f7fe8ff8be8c5f53a5/geofm_src/engine/accelerated/utils/metrics.py#L75).

Again, thank you for your time investment to catch this bug! Once I have double checked out new with your results, I will update the values with corresponding attribution to you.

One follow-up about geobreeze: I know there are a lot of other evaluation codebases out there. From your experience, is geobreeze useful & filling a gap? If not, what could we improve? No hard feelings, just interested in improvement! Also happy to chat via mail if you prefer leonard.waldmann@tum.de

**Comment by LeWaldm** [comment]
Hi @JohnMBrandt ,

My apologies for not coming back to you earlier. Here my analysis.


## Metric definitions
I did some digging and in general, macro jaccard is mIoU, but to obtain mIoU per class over multiple images, there are 2 options: (1) compute mIoU over each image and then average the scores over multiple images (torchmetrics.segmentation.MeanIoU) or (2) count intersection and union pixels over all images and do a single division in the end (torchmetrics.classification.JaccardIndex(average=macro)). So while micro jaccard (which we reported in the paper) is definitely not miou, miou over multiple images is not universally defined (c.f. https://stats.stackexchange.com/questions/554724/is-there-an-official-procedure-to-compute-miou-mean-intersection-over-union). This might cause additional discrepancies in reported mIoU scores.



## Results
We re-evaluated the experiment with micro Jaccard (which we reported in the paper), macro Jaccard (mIoU aggregating counts over multiple images), and torchmetrics.segmentation.MeanIoU (averaged over images). Empty values could not be computed due to band mismatches.

### Jaccard micro

| model      | m-cashew-plant   | m-chesapeake   | m-NeonTree   | m-nz-cattle   | m-pv4ger-seg   | m-SA-crop-type   |
|------------|------------------|----------------|--------------|---------------|----------------|------------------|
| dinov2     | **68.1**         | **77.6**       | **83.3**     | **97.3**      | **97.1**       | 50.7             |
| croma      | 54.2             |                |              |               |                | 51.1             |
| softcon    | 57.1             |                |              |               |                | 51.4             |
| anysat     | 42.5             | 75.9           | 80.9         | 96.4          | 93.4           | 44.4             |
| galileo    | 50.2             |                |              |               |                | 47.4             |
| senpamae   | 39.9             | 68.0           | 80.6         | 95.4          | 88.0           | 41.4             |
| dofa       | 58.8             | 77.4           | 82.3         | 97.0          | 96.8           | 50.7             |
| panopticon | 62.3             | **77.6**       | 80.6         | 97.1          | 96.5           | **53.1**         |



### Jaccard macro
| model      | m-cashew-plant   | m-chesapeake   | m-NeonTree   | m-nz-cattle   | m-pv4ger-seg   | m-SA-crop-type   |
|------------|------------------|----------------|--------------|---------------|----------------|------------------|
| dinov2     | **53.4**         | **64.0**       | **59.1**     | **76.9**      | **95.2**       | 32.0             |
| croma      | 34.3             |                |              |               |                | 34.4             |
| softcon    | 39.9             |                |              |               |                | 33.7             |
| anysat     | 21.8             | 61.7           | 49.9         | 67.6          | 89.2           | 23.6             |
| galileo    | 30.4             |                |              |               |                | 28.4             |
| senpamae   | 17.8             | 46.9           | 48.3         | 49.6          | 80.4           | 18.4             |
| dofa       | 41.9             | 59.2           | 55.4         | 74.3          | 94.6           | 32.5             |
| panopticon | 45.1             | 60.8           | 50.4         | 73.9          | 94.1           | **35.7**         |





### MeanIoU 
| model      | m-cashew-plant   | m-chesapeake   | m-NeonTree   | m-nz-cattle   | m-pv4ger-seg   | m-SA-crop-type   |
|------------|------------------|----------------|--------------|---------------|----------------|------------------|
| dinov2     | **52.7**         | **26.8**       | **55.8**     | **77.4**      | **92.7**       | 13.9             |
| croma      | 33.8             |                |              |               |                | 14.2             |
| softcon    | 39.2             |                |              |               |                | 14.2             |
| anysat     | 20.9             | 25.5           | 47.4         | 67.9          | 86.0           | 10.0             |
| galileo    | 30.1             |                |              |               |                | 11.9             |
| senpamae   | 17.6             | 19.7           | 46.3         | 49.0          | 69.4           | 9.1              |
| dofa       | 41.2             | 26.6           | 51.5         | 74.6          | 89.7           | 13.9             |
| panopticon | 44.7             | 26.6           | 47.8         | 73.0          | 88.2           | **14.9**         |


## Interpretation
### Rankings
We compare the rankings of the models with the different metrics using spearman and kendall. We obtain high values close to 1, indicating that the rankings are very similar. 

|                     |                           |   m-NeonTree |   m-SA-crop-type |   m-cashew-plant |   m-chesapeake |   m-nz-cattle |   m-pv4ger-seg |
|:--------------------|---------------------------|-------------:|-----------------:|-----------------:|---------------:|--------------:|---------------:|
| jaccard_macro vs jaccard_micro | Spearman |         0.82 |             0.97 |                1 |           0.62 |           0.9 |              1 |
|  | Kendall  |         0.74 |             0.91 |                1 |           0.53 |           0.8 |              1 |
| jaccard_macro vs mIoU | Spearman          |         1    |             0.99 |                1 |           0.67 |           1   |              1 |
|  | Kendall         |         1    |             0.96 |                1 |           0.53 |           1   |              1 |
| jaccard_micro vs mIoU  | Spearman          |         0.82 |             0.99 |                1 |           0.92 |           0.9 |              1 |
|  | Kendall         |         0.74 |             0.98 |                1 |           0.89 |           0.8 |              1 |

### Magnitudes
However, the magnitudes change substantially. 
- micro jaccard seems to be strictly higher than macro jaccard, especially on neontree, cattle, and sa-crop-type, all of which have a frequent background class (see below, half of the bar on very left and right are cutoff). Over-predicting the background class in these cases is not punished hard in micro jaccard. 
<img width="791" height="490" alt="Image" src="https://github.com/user-attachments/assets/2196cb44-3af2-43fe-a9bd-f55d813ddc26" />

- MIoU seems to be slightly lower than macro jaccard and substantially lower on m-chesapeak and m-sa-crop-type. One possible explanation is that there are many occurrences of only few pixels belonging to a class, which have more weight in miou if predicted falsely. Indeed, plotting the cumulative occurrences of number of pixels per class over all images (removing count=0 occurrences), we observe many occurrences of classes with few pixels in chesapeak and crop-type. We note that nz-cattle also has many observations of few pixels in a class but macro jaccard and miou are very similar. I am not entirely sure why this happens, but I could imagine the different number of classes (2 nz-cattle, 7 chesapeak, 10 m-sa-crop-type) or potentially nz-cattle being easier to predict to be a factor.
<img width="690" height="390" alt="Image" src="https://github.com/user-attachments/assets/94c6a1da-76f2-4dfd-b79f-9496cb4f0747" />

## Discussion
Overall, while the metrics have different magnitudes, the relative ordering of the models remain similarly and thus, the conclusions of the paper remain intact.

We will put the tables with the different metrics in an updated version on arxiv end of next week. Thank you again, for pointing out this bug @JohnMBrandt ! If you have any further questions, let us know.
